spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/store/jwk/JwkDefinitionSource.java

```diff
@@ -96,10 +96,12 @@ JwkDefinitionHolder getDefinitionLoadIfNecessary(String keyId) {
 			if (result != null) {
 				return result;
 			}
-			this.jwkDefinitions.clear();
+			Map newJwkDefinitions = new LinkedHashMap<String, JwkDefinitionHolder>();
 			for (URL jwkSetUrl : jwkSetUrls) {
-				this.jwkDefinitions.putAll(loadJwkDefinitions(jwkSetUrl));
+				newJwkDefinitions.putAll(loadJwkDefinitions(jwkSetUrl));
 			}
+			this.jwkDefinitions.clear();
```

Comment:
@JanGroot Reorganizing the `clear()` just before the `putAll()` won't do anything to improve the performance here since `synchronized (this.jwkDefinitions)` will block until all JWK's have been fetched and stored. This will need to be re-worked in order to improve performance by reducing the blocking time.

Reply:
Hi, thanks for looking into this.

The point of moving the clear to after the *loadJwkDefinitions* is to make sure the other threads with a correct KID never reach the synchronized block as they will still return on line 91 as the keys are not cleared yet and they will still be found.

The only threads waiting on the fetch will be those with an unknown or bogus key.

OLD situation:
definitions contains A
t1 enters with key B -> definitions is cleared -> start fetching keys
t2, t3, t4 , t5 enter with key A -> definition not found -> all block on fetch
t1 -> put keys (only A as B does not exist) -> returns null 
t2 returns key A

New 
definitions contains A
t1 enters with key B -> block on  start fetching keys
t2, t3, t4, t5, t6  enter with key A -> returns key A
t1 -> clears key -> puts keys -> returns null for B

Reply:
Just seeing now this project is in maintenance, in spring-sec 5 this is ok already via
`com.nimbusds.jose.jwk.source.RemoteJWKSet#updateJWKSetFromURL`
there the cache is never cleared to begin with.

Reply:
@JanGroot Yes, you are correct. Moving the `clear()`  and `putAll()` after the retrieval will help. I missed this in my initial analysis.